clippy_lints/src/partialeq_ne_impl.rs

```diff
@@ -0,0 +1,55 @@
+use rustc::lint::*;
+use rustc::hir::*;
+use utils::{is_automatically_derived, span_lint};
+
+/// **What it does:** Checks for manual re-implementations of `PartialEq::ne`.
+///
+/// **Why is this bad?** `PartialEq::ne` is required to always return the
+/// negated result of `PartialEq::eq`, which is exactly what the default
+/// implementation does. Therefore, there should never be any need to
+/// re-implement it.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// struct Foo;
+///
+/// impl PartialEq for Foo {
+///    fn eq(&self, other: &Foo) -> bool { ... }
+///    fn ne(&self, other: &Foo) -> bool { !(self == other) }
+/// }
+/// ```
+declare_lint! {
+    pub PARTIALEQ_NE_IMPL,
+    Warn,
+    "re-implementing PartialEq::ne"
+}
+
+#[derive(Clone, Copy)]
+pub struct Pass;
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(PARTIALEQ_NE_IMPL)
+    }
+}
+
+impl LateLintPass for Pass {
+    fn check_item(&mut self, cx: &LateContext, item: &Item) {
+        if_let_chain! {[
+            let ItemImpl(_, _, _, Some(ref trait_ref), _, ref impl_items) = item.node,
+            !is_automatically_derived(&*item.attrs),
+            cx.tcx.expect_def(trait_ref.ref_id).def_id() == cx.tcx.lang_items.eq_trait().unwrap(),
+        ], {
+            for impl_item in impl_items {
+                if &*impl_item.name.as_str() == "ne" {
+                    span_lint(cx,
+                              PARTIALEQ_NE_IMPL,
+                              impl_item.span,
+                              "Re-implementing `PartialEq::ne`")
+                }
```

Comment:
This should also give a help message telling the user what to do( remove the impl)